bigchaindb/models.py

```diff
@@ -0,0 +1,209 @@
+            DoubleSpend: if the transaction is a double spend
+            InvalidHash: if the hash of the transaction is wrong
+            InvalidSignature: if the signature of the transaction is wrong
+        """
+        if len(self.fulfillments) == 0:
+            raise ValueError('Transaction contains no fulfillments')
+
+        input_conditions = []
+        inputs_defined = reduce(and_, [bool(ffill.tx_input) for ffill
+                                       in self.fulfillments])
+        if self.operation in (Transaction.CREATE, Transaction.GENESIS):
+            if inputs_defined:
+                raise ValueError('A CREATE operation has no inputs')
+        elif self.operation == Transaction.TRANSFER:
+            if not inputs_defined:
+                raise ValueError('Only `CREATE` transactions can have null inputs')
+
+            for ffill in self.fulfillments:
+                input_txid = ffill.tx_input.txid
+                input_cid = ffill.tx_input.cid
+                input_tx = bigchain.get_transaction(input_txid)
+                if input_tx is None:
+                    raise TransactionDoesNotExist("input `{}` doesn't exist"
+                                                  .format(input_txid))
+                else:
+                    input_conditions.append(input_tx.conditions[input_cid])
+
+                spent = bigchain.get_spent(input_txid, ffill.tx_input.cid)
+                if spent and spent.id != self.id:
+                    raise DoubleSpend('input `{}` was already spent'
+                                      .format(input_txid))
+        else:
+            raise TypeError('`operation` must be either `TRANSFER`, `CREATE` '
+                            'or `GENESIS`')
+
+        if not self.fulfillments_valid(input_conditions):
+            raise InvalidSignature()
+        else:
+            return self
+
+
+class Block(object):
+    def __init__(self, transactions=None, node_pubkey=None, timestamp=None,
+                 voters=None, signature=None):
+        if transactions is not None and not isinstance(transactions, list):
+            raise TypeError('`transactions` must be a list instance or None')
+        elif transactions is None:
+            self.transactions = []
+        else:
+            self.transactions = transactions
+
+        if voters is not None and not isinstance(voters, list):
+            raise TypeError('`voters` must be a list instance or None')
+        elif transactions is None:
+            self.voters = []
+        else:
+            self.voters = voters
+
+        if timestamp is not None:
+            self.timestamp = timestamp
+        else:
+            self.timestamp = gen_timestamp()
+
+        self.node_pubkey = node_pubkey
+        self.signature = signature
+
+    def __eq__(self, other):
+        try:
+            other = other.to_dict()
+        except AttributeError:
+            return False
+        return self.to_dict() == other
```

Comment:
> For instance, it is not clear to me why an empty block cannot have a dictionary representation. It could be an empty dict {}, perhaps, or something like:

That's absolutely right. Usually we validate the block right after serialization anyways.